book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,778 @@
+    - `(height, block)` into `block_by_height`.
+
+3. If the block is a genesis block, skip any transaction updates.
+
+(Due to a [bug in zcashd](https://github.com/ZcashFoundation/zebra/issues/559), genesis block transactions
+are ignored during validation.)
+
+3. If the block is a genesis block, skip any transaction updates.
+
+(Due to a [bug in zcashd](https://github.com/ZcashFoundation/zebra/issues/559), genesis block transactions
+are ignored during validation.)
+
+3.  Update the `sprout_anchors` and `sapling_anchors` trees with the Sprout and Sapling anchors.
+
+4. Iterate over the enumerated transactions in the block. For each transaction:
+
+   1. Insert `(transaction_hash, block_height || BE32(tx_index))` to
+   `tx_by_hash`;
+
+   2. For each `TransparentInput::PrevOut { outpoint, .. }` in the
+   transaction's `inputs()`, remove `outpoint` from `utxo_by_output`.
+
+   3. For each `output` in the transaction's `outputs()`, construct the
+   `outpoint` that identifies it, and insert `(outpoint, output)` into
+   `utxo_by_output`.
+
+   4. For each [`JoinSplit`] description in the transaction,
+   insert `(nullifiers[0],())` and `(nullifiers[1],())` into
+   `sprout_nullifiers`.
+
+   5. For each [`Spend`] description in the transaction, insert
+   `(nullifier,())` into `sapling_nullifiers`.
+
+**Note**: The Sprout and Sapling anchors are the roots of the Sprout and
+Sapling note commitment trees that have already been calculated for the last
+transaction(s) in the block that have `JoinSplit`s in the Sprout case and/or
+`Spend`/`Output` descriptions in the Sapling case. These should be passed as
+fields in the `Commit*Block` requests.
+
+[`JoinSplit`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.JoinSplit.html
+[`Spend`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.Spend.html
+
+These updates can be performed in a batch or without necessarily iterating
+over all transactions, if the data is available by other means; they're
+specified this way for clarity.
+
+
+## Request / Response API
+[request-response]: #request-response
+
+The state API is provided by a pair of `Request`/`Response` enums. Each
+`Request` variant corresponds to particular `Response` variants, and it's
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock`
+[request-commit-block]: #request-commit-block
+
+```rust
+CommitBlock {
+    block: Arc<Block>,
+    sprout_anchor: sprout::tree::Root,
+    sapling_anchor: sapling::tree::Root,
```

Comment:
@dconnolly does every post-Sapling block have a sprout and a sapling anchor?

Reply:
Depends on the contents of the transactions in the block, joinsplits for sprout and spends/outputs for sapling.

 If there are no joinsplits, then the input sprout treestate from the parent block has not been changed, and I _think_ (but starting at the spec to confirm) that means that the output Sprout treestate == the input Sprout treestate. Same for Sapling, if there are no changes to the state, I assume the final treestate (anchor) is the output state from the last transaction (a no-op).

Reply:
Ok, that sounds like we'll always have both `Root`s in CommitBlock, because it only accepts post-Sapling blocks.

But I just realised that `CommitFinalizedBlock` accepts all blocks, including Genesis blocks.

Do genesis blocks have a Sprout root?
Do pre-Sapling blocks have a Sapling root?

Reply:
For the genesis block, the input treestate 'for each of Sprout and Sapling', is the empty treestate, which we can compute with my changes in #1045  (at least for Sapling for now, I'm still working on Sprout).

So for pre-Sapling, I think they literally compute and pass along the empty Sapling treestate, from block to block, until Sapling activation.

![image](https://user-images.githubusercontent.com/552961/94092249-3a440e00-fde8-11ea-937d-855b96f22b44.png)

Reply:
Oh, and by definition, the genesis block can't have a Sprout transaction, because there are no coins yet.

Ok, that's great, I think we're good, as far as the state goes.

I'll leave this conversation un-resolved, so @str4d or @daira can double-check things.

Reply:
Technically there is an unspendable block subsidy in the genesis block (as with a Bitcoin); it just happens that due to the slow start, it has a value of zero. In a chain where ZIP 213 is active from genesis, there could therefore be a Sapling output in the genesis block; if that chain also disabled slow start, the output would have value. It's not immediately clear to me at 4am whether that output should be spendable; under the current consensus rules, I think it would be added to the commitment tree and therefore be spendable (whereas genesis transparent UTXOs are not added to the UTXO set).

In practice the genesis block never contains shielded outputs in `zcashd` due to the way that the genesis block is side loaded, making it hard to generate on the fly and thus simply hard-coded for regtest. In our RPC tests we currently activate the NUs up to Sapling at height 1.

Reply:
Also, transparent coinbase UTXOs can't be spent until 100 blocks after they are created :-)